modules/storage/src/StockStorageAPI.php

```diff
@@ -21,180 +21,200 @@ public function createTransaction($variation_id, $location_id, $zone, $quantity,
-      $related_tid,
-      $related_oid,
-      $related_uid,
-      serialize($data),
-    ]);
-    return $io->execute();
+    $field_values = [
+      'variation_id' => $variation_id,
+      'qty' => $quantity,
+      'location_id' => $location_id,
+      'location_zone' => $zone,
+      'unit_cost' => $unit_cost,
+      'transaction_time' => time(),
+      'transaction_type_id' => $transaction_type_id,
+      'related_tid' => $related_tid,
+      'related_oid' => $related_oid,
+      'related_uid' => $related_uid,
+      'data' => serialize($data),
+    ];
+    $insert = \Drupal::database()->insert('cs_inventory_transaction')
+      ->fields(array_keys($field_values))
+      ->values(array_values($field_values))->execute();
+
+    return $insert;
   }
 
   /**
    * {@inheritdoc}
    */
-  public function updateProductInventoryLocationLevel($location_id, $variation_id) {
-    // Get the location level & last transaction.
+  public function updateLocationStockLevel($location_id, $variation_id) {
+    $current_level = $this->getLocationStockLevel($location_id, $variation_id);
+    $last_update = $current_level['last_transaction'];
+    $latest_txn = $this->getLocationStockTransactionLatest($location_id, $variation_id);
+    $latest_sum = $this->getLocationStockTransactionSum($location_id, $variation_id, $last_update, $latest_txn);
+    $new_level = $current_level['qty'] + $latest_sum;
+
+    $this->setLocationStockLevel($location_id, $variation_id, $new_level, $latest_txn);
+  }
+
+  /**
+   * Gets stock level for a given location and variation.
+   *
+   * @param $location_id
+   *   Location id.
+   *
+   * @param $variation_id
+   *   Variation id.
+   *
+   * @return array
+   *   An array of 'qty' and 'last_transaction_id' values.
+   */
+  public function getLocationStockLevel($location_id, $variation_id) {
     $db = \Drupal::database();
     $result = $db->select('cs_inventory_location_level', 'ill')
       ->fields('ill')
-      ->condition('location_id', $location_id, '=')
+      ->condition('location_id', $location_id)
       ->condition('variation_id', $variation_id)
       ->execute()
       ->fetch();
-    if ($result) {
-      $stock_level = $result->qty;
-      $last_transaction = $result->last_transaction_id;
+
+    return [
+      'qty' => $result ? $result->qty : 0,
+      'last_transaction' => $result ? $result->last_transaction_id : 0
+    ];
+  }
+
+  /**
+   * Sets the stock level and last transaction for a given location and variation.
+   *
+   * Creates first stock level transaction record if none exists.
+   *
+   * @param int $location_id
+   *   Location id.
+   *
+   * @param int $variation_id
+   *   Variation id.
+   *
+   * @param int $qty
+   *   Quantity.
+   *
+   * @param int $last_txn
+   *   Last transaction id.
+   */
+  public function setLocationStockLevel($location_id, $variation_id, $qty, $last_txn) {
+    $db = \Drupal::database();
+    $existing = $db->select('cs_inventory_location_level', 'ill')
```

Comment:
instead of testing for existence, I think you can use $db->merge() to handle update/insert